h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java

```diff
@@ -1,32 +1,163 @@
+                                        final int nodeIndex, final SharedTreeSubgraph sharedTreeSubgraph,
+                                        final FeatureProperties featureProperties, boolean inclusiveNA) {
+    final RegTreeNode xgBoostNode = xgBoostNodes[nodeIndex];
+    // Not testing for NaNs, as SharedTreeNode uses NaNs as default values.
+    //No domain set, as the structure mimics XGBoost's tree, which is numeric-only
+    if (featureProperties._oneHotEncoded[xgBoostNode.split_index()]) {
+      //Shared tree model uses < to the left and >= to the right. Transforiming one-hot encoded categoricals
+      // from 0 to 1 makes it fit the current split description logic
+      sharedTreeNode.setSplitValue(1.0F);
+    } else {
+      sharedTreeNode.setSplitValue(xgBoostNode.getSplitCondition());
+    }
+    sharedTreeNode.setPredValue(xgBoostNode.getLeafValue());
+    sharedTreeNode.setCol(xgBoostNode.split_index(), featureProperties._names[xgBoostNode.split_index()]);
+    sharedTreeNode.setInclusiveNa(inclusiveNA);
+    sharedTreeNode.setNodeNumber(nodeIndex);
+
+    if (xgBoostNode.getLeftChildIndex() != -1) {
+      constructSubgraph(xgBoostNodes, sharedTreeSubgraph.makeLeftChildNode(sharedTreeNode),
+              xgBoostNode.getLeftChildIndex(), sharedTreeSubgraph, featureProperties, xgBoostNode.default_left());
+    }
+
+    if (xgBoostNode.getRightChildIndex() != -1) {
+      constructSubgraph(xgBoostNodes, sharedTreeSubgraph.makeRightChildNode(sharedTreeNode),
+              xgBoostNode.getRightChildIndex(), sharedTreeSubgraph, featureProperties, !xgBoostNode.default_left());
+    }
+  }
+
+
+  private static FeatureProperties assembleFeatureNames(final DataInfo di) {
+    String[] coefnames = di.coefNames();
+    assert (coefnames.length == di.fullN());
+    int numCatCols = di._catOffsets[di._catOffsets.length - 1];
+
+    String[] featureNames = new String[di.fullN()];
+    boolean[] oneHotEncoded = new boolean[di.fullN()];
+    for (int i = 0; i < di.fullN(); ++i) {
+      featureNames[i] = coefnames[i];
+      if (i < numCatCols) {
+        oneHotEncoded[i] = true;
+      }
+    }
+
+    return new FeatureProperties(featureNames, oneHotEncoded);
+  }
+
+  private static class FeatureProperties {
+    private String[] _names;
+    private boolean[] _oneHotEncoded;
+
+    public FeatureProperties(String[] names, boolean[] oneHotEncoded) {
+      _names = names;
+      _oneHotEncoded = oneHotEncoded;
+    }
+  }
+
+  private final int getXGBoostClassIndex(final String treeClass) {
+    final ModelCategory modelCategory = _output.getModelCategory();
+    if (treeClass == null && ModelCategory.Regression.equals(modelCategory)) return 0;
+    if (treeClass == null && !ModelCategory.Regression.equals(modelCategory)) {
+      throw new IllegalArgumentException("Non-regressional models require tree class specified.");
+    }
+
+    final String[] domain = Objects.requireNonNull(model_info._dataInfoKey.get())
```

Comment:
you can use _output._domains - response will be always the last one

Reply:
Good catch ! Thanks. Speeds things up a little bit as well :)